tritium-processor/src/main/java/com/palantir/tritium/processor/ProxyAnnotationProcessor.java

```diff
@@ -0,0 +1,426 @@
+        if (mirror.getKind() == TypeKind.ERROR) {
+            return true;
+        }
+        List<DeclaredType> minimalInterfaces = minimalInterfacesFinal;
+        if (mirror.getKind() == TypeKind.DECLARED && types.asElement(mirror).getKind() == ElementKind.INTERFACE) {
+            allInterfaces.add((DeclaredType) mirror);
+            if (minimalInterfaces != null) {
+                minimalInterfaces.add((DeclaredType) mirror);
+                minimalInterfaces = null;
+            }
+        }
+        return isInvalid(((TypeElement) types.asElement(mirror)).getInterfaces(), allInterfaces, minimalInterfaces);
+    }
+
+    @SuppressWarnings("checkstyle:CyclomaticComplexity")
+    private JavaFile generate(
+            TypeElement typeElement, List<DeclaredType> allInterfaces, List<DeclaredType> minimalInterfaces) {
+        TypeName annotatedType = TypeName.get(typeElement.asType());
+        String packageName =
+                elements.getPackageOf(typeElement).getQualifiedName().toString();
+        String className = "Proxied" + typeElement.getSimpleName();
+        List<TypeName> interfaceNames = new ArrayList<>(minimalInterfaces.size());
+        List<TypeVariableName> typeVarNames = new ArrayList<>();
+        for (DeclaredType type : minimalInterfaces) {
+            interfaceNames.add(TypeName.get(type));
+            for (TypeMirror typeArg : type.getTypeArguments()) {
+                if (typeArg instanceof TypeVariable) {
+                    typeVarNames.add(TypeVariableName.get((TypeVariable) typeArg));
+                }
+            }
+        }
+
+        TypeSpec.Builder specBuilder = TypeSpec.classBuilder(className)
+                .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
+                .addOriginatingElement(typeElement);
+
+        GeneratedAnnotations.generatedAnnotation(elements, SourceVersion.latest())
+                .ifPresent(te -> specBuilder.addAnnotation(AnnotationSpec.builder(ClassName.get(te))
+                        .addMember("value", "$S", getClass().getName())
+                        .build()));
+
+        if (typeElement.getAnnotation(Deprecated.class) != null) {
+            specBuilder.addAnnotation(Deprecated.class);
+        }
+
+        specBuilder
+                .addSuperinterfaces(interfaceNames)
+                .addTypeVariables(typeVarNames)
+                .addField(FieldSpec.builder(InvocationHandler.class, HANDLER_NAME, Modifier.PRIVATE, Modifier.FINAL)
+                        .build())
+                .addMethod(MethodSpec.constructorBuilder()
+                        .addModifiers(Modifier.PRIVATE)
+                        .addParameter(ParameterSpec.builder(InvocationHandler.class, HANDLER_NAME)
+                                .build())
+                        .addStatement("this.$1N = $1N", HANDLER_NAME)
+                        .build());
+
+        Map<String, List<MethodElements>> methodsByName = new HashMap<>();
```

Comment:
Without going through the entirety of this logic, would MoreElements#getAllMethods work?

Reply:
I've replaced the visitor with `MoreElements.getLocalAndInheritedMethods`, however that method can return duplicates if they're inherited from two separate superinterfaces, so we still need some of the complexity.